### PR TITLE
Renable e2e test with automated hosted zone provisioning

### PR DESCRIPTION
**Pull request by arjunrn**

### CODE REVIEW

File: test/e2e/operator_test.go [hunk 1]
```diff
@@ -62,240 +49,185 @@ func init() {
...
-			os.Exit(1)
-		}
-		defaultParentZoneID = os.Getenv("EXTDNS_PARENT_ZONEID")
-		if defaultParentZoneID == "" {
-			fmt.Printf("No parent zone ID specified. Please set the environment variable EXTDNS_PARENT_ZONEID\n")
-			os.Exit(1)
-		}
-
-		domainName = generateDomainName()
+	hostedZoneID, nameServers, err = helper.ensureHostedZone(hostedZoneDomain)
```
Comment (alebedev87) [inline test/e2e/operator_test.go:122]: If there are 2 PR running this e2e test, won't they race on the usage on the same hostedZone? Knowing also that the test service has the hard coded name, there can be false positive and negative tests.
This may need to be refactored in the followups to try to get the unique hostedzone name with the commit checksum in the name, for instance.

File: test/e2e/operator_test.go [hunk 2]
```diff
@@ -62,240 +49,185 @@ func init() {
...
-		httpClient = &http.Client{
-			Transport: &http.Transport{
-				Dial:        dialer.Dial,
-				DialContext: dialer.DialContext,
-			},
-		}
-
-		os.Exit(m.Run())
-	*/
+	err = helper.deleteHostedZone(hostedZoneID)
```
Comment (alebedev87) [inline test/e2e/operator_test.go:130]: Do we want to `defer` this deletion right after successful `ensureHostedZone`?

File: go.mod [hunk 3]
```diff
@@ -9,6 +9,7 @@ require (
 	github.com/onsi/ginkgo v1.16.4
 	github.com/onsi/gomega v1.15.0
 	github.com/openshift/api v0.0.0-20211019100638-b2cbe79f2e4b
+	github.com/openshift/cloud-credential-operator v0.0.0-20211115195328-4c537fbd48e9
```
Comment (alebedev87) [inline go.mod:12]: Do we still need this dependency knowing that we're getting the root credentials now?

File: test/e2e/operator_test.go (hunk 2 above)
Comment (arjunrn) [reply test/e2e/operator_test.go:130]: We cannot deferred calls because we have to call `os.Exit()` and that immediately exits the program. We can move all the logic into another function just to use defer but I think that might be overkill for now.

File: test/e2e/operator_test.go (hunk 1 above)
Comment (arjunrn) [reply test/e2e/operator_test.go:122]: Yes, this need to be done otherwise we might get too many flakes. Let's not merge this as-is.

File: test/e2e/operator_test.go [hunk 4]
```diff
@@ -6,53 +6,40 @@ import (
...
 )
 
 const (
-	recordExistsTimeout  = 5 * time.Minute
-	podReadyTimeout      = 1 * time.Minute
-	httpGetTimeout       = 10 * time.Minute
-	recordDeletedTimeout = 5 * time.Minute
-	dnsLookupTimeout     = 5 * time.Second
-	dnsServer            = "8.8.8.8:53"
+	hostedZoneDomain   = "example-test.info"
```
Comment (arjunrn) [inline test/e2e/operator_test.go:30]: The hosted zone need to be randomized so concurrent executions of the e2e tests do not interfere with each other.

File: test/e2e/operator_test.go (hunk 2 above)
Comment (alebedev87) [reply test/e2e/operator_test.go:127]: Yes, it's kind of a choice between:
- last step cleanup (current implementation) which fails the test if the cleanup broke but leave leftover zones of the test cases panic
- best effort cleanup (no `Exit` with `defer`) which won't fail the test if the clean broke but will try to do something the test cases panic

File: test/e2e/operator_test.go [hunk 5]
```diff
@@ -64,238 +51,180 @@ func init() {
...
-		current := podConditionMap(pod.Status.Conditions...)
-		return conditionsMatchExpected(expected, current), nil
-	})
-	err = wait.PollImmediate(5*time.Second, httpGetTimeout, func() (bool, error) {
-		var resp *http.Response
-		resp, err := httpClient.Get(fmt.Sprintf("http://%s", expectedDomainName))
+
+		// get the IPs of the loadbalancer
+		lbHostname := service.Status.LoadBalancer.Ingress[0].Hostname
+		ips, err := net.LookupIP(lbHostname)
```
Comment (arjunrn) [inline test/e2e/operator_test.go:189]: This uses the default resolver where `PreferGo` is false.

**Comment by alebedev87** [comment]
Merging this one to iterate on the remark in later PRs.
/lgtm
